xmhf/src/xmhf-core/xmhf-runtime/xmhf-smpguest/arch/x86/vmx/smpg-x86vmx.c

```diff
@@ -653,83 +653,172 @@ void xmhf_smpguest_arch_x86vmx_eventhandler_nmiexception(VCPU *vcpu, struct regs
-        // Thus, "vcpu->vmcs.control_VMX_cpu_based" can be seen as a NMI pending bit of the current guest.
-
-        // [NOTE] Updating "vcpu->vmcs.control_VMX_cpu_based" here without "vcpu->vmx_guest_inject_nmi = 1" cannot 
-        // guarantee control_VMX_cpu_based is written. This is because VMexit handlers have 
-        // xmhf_baseplatform_arch_x86vmx_getVMCS and xmhf_baseplatform_arch_x86vmx_putVMCS or similar pair of VMCS 
-        // backup/restore operations, and they form race conditions with any "vcpu->vmcs.control_VMX_cpu_based" writing
-        // in this function.
-
-		/* [Detailed Explanations]
+		/*
+		 * Note: injecting NMI into the guest OS need to be done with extra
+		 * carefulness. There are 2 challenges present.
+		 *
+		 * First, if the guest is running NMI handler and has not executed the
+		 * IRET instruction, injecting NMI to the guest will corrupt the guest.
+		 * So the VMX root must not use
+		 * vcpu->vmcs.control_VM_entry_interruption_information to directly
+		 * inject NMI into the current guest. Instead, the hypervisor should
+		 * use the "NMI-window exiting" VM-execution control to be notified
+		 * when the guest is able to handle NMI interrupts. When the guest can
+		 * handle NMI interrupts, an VMEXIT will occur with reason "NMI
+		 * window". This requires "NMI exiting" and "virtual NMIs" bits to be
+		 * set in Pin-Based VM-Execution Controls.
+		 *
+		 * Second, modifying VMCS is subject to race conditions. During normal
+		 * operations, the hypervisor does not block NMIs. It is also
+		 * difficult, if not impossible, to let the hypervisor block NMI
+		 * actively. So when writing NMI-related code, all interleavings
+		 * between the code and the NMI interrupt handler need to be
+		 * considered. It may be helpful to think the NMI interrupt handler as
+		 * a signal handler of a process in UNIX-like systems.
+		 *
+		 * To solve the second problem, we introduce a variable in vcpu to
+		 * denote that the NMI interrupt handler has visited and wants to set
+		 * the "NMI-window exiting" bit (22nd bit in control_VMX_cpu_based).
+		 * This variable is called vmx_guest_vmcs_nmi_window_set. Before the
+		 * intercept handler returns, it checks vmx_guest_vmcs_nmi_window_set.
+		 * If the variable is true, the "NMI-window exiting" bit in VMCS is
+		 * set.
+		 *
+		 * Also note that vcpu->vmcs.control_VMX_cpu_based does not access
+		 * the real VMCS in hardware. The real VMCS can only be accessed
+		 * through VMREAD and VMWRITE instructions. vcpu->vmcs is loaded /
+		 * stored at begin / end of intercepts using
+		 * xmhf_baseplatform_arch_x86vmx_getVMCS() and
+		 * xmhf_baseplatform_arch_x86vmx_putVMCS().
+		 *
+		 * [Detailed explanation for the second problem follows]
+		 *
 		 * We want to set vcpu->vmcs.control_VMX_cpu_based, but we may be in an
-		 * exception handler that is interrupting an intercept handler using
```

Comment:
It should be better to modify all "intercept handler" to be "VMExit handler", and replace "NMI exception handler" to be "NMI interrupt handler"